docs/rfc/rfc-002-ipc-ecosystem.md

```diff
@@ -0,0 +1,348 @@
+
+TODO: It's not clear to me how meaningfully a remote signer is really isolated
+by this mechanism. The consensus node has to be on a shared network with the
+remote signer, so a compromise of the consensus node from the public network
+leaves the attacker able to communicate with the remote signer anyway. It does
+raise the bar by a constant factor -- now you also have to compromise the
+remote signer too -- but it's not actually isolated in the gRPC configuration.
+
+
+### ABCI Transport
+
+In ABCI, the _application_ is the server, and the Tendermint consensus engine
+is the client.  Most applications implement the server using the [Cosmos SDK][cosmos-sdk],
+which handles low-level details of the ABCI interaction and provides a
+higher-level interface to the rest of the application. The SDK is written in Go.
+
+Beneath the SDK, the application communicates with Tendermint core in one of
+two ways:
+
+- In-process direct calls (for applications written in Go and compiled against
+  the Tendermint code).  This is an optimization for the common case where an
+  application is written in Go, to save on the overhead of marshaling and
+  unmarshaling requests and responses within the same process:
+  [`abci/client/local_client.go`][local-client]
+
+- A custom remote procedure protocol built on wire-format protobuf messages
+  using a socket (the "socket protocol"): [`abci/server/socket_server.go`][socket-server]
+
+The SDK also provides a [gRPC service][sdk-grpc] accessible from outside the
+application, allowing transactions to be broadcast to the network, look up
+transactions, and simulate transaction costs.
+
+
+### RPC Transport
+
+The consensus node RPC service allows callers to query consensus parameters
+(genesis data, transactions, commits), node status (network info, health
+checks), application state (abci_query, abci_info), mempool state, and other
+attributes of the node and its application. The service also provides methods
+allowing transactions and evidence to be injected ("broadcast") into the
+blockchain.
+
+The RPC service is exposed in several ways:
+
+- HTTP GET: Queries may be sent as URI parameters, with method names in the path.
+
+- HTTP POST: Queries may be sent as JSON-RPC request messages in the body of an
+  HTTP POST request.  The server uses a custom implementation of JSON-RPC that
+  is not fully compatible with the [JSON-RPC 2.0 spec][json-rpc], but handles
+  the common cases.
+
+- Websocket: Queries may be sent as JSON-RPC request messages via a websocket.
+  This transport uses more or less the same JSON-RPC plumbing as the HTTP POST
+  handler.
+
+  TODO: Three methods, maybe related to event subscription, are _only_ exported
+  via websocket. I think these may be used by the light client for
+  synchronization. In any case, the light client appears to be the only direct
+  consumer of the websocket transport at least within the core repository.
+  Figure out whether that's true.
```

Comment:
Does the core light client implementation actually use the event system though? AFAIK it has nothing to do with it. Only with the IBC case with relayers do we see events used to update the light client. Perhaps in the future it may be wise to natively support this case but as for now I think they are separate

Reply:
Good question—I will have to do a bit more investigation, and will update this once I have. 🙂

Reply:
The relayer does use event subscription, and also uses the `ABCIQuery` method. As far as I can see, it does not use any other of the RPC service endpoints.